### PR TITLE
KTIJ-5953 Add quickfix to 'create companion object and member\extension function'

### PR DESCRIPTION
**Pull request by t-kameyama**
https://youtrack.jetbrains.com/issue/KTIJ-5953

### CODE REVIEW

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createCallable/CreateCallableFromCallActionFactory.kt [hunk 1]
```diff
@@ -45,6 +45,7 @@ sealed class CreateCallableFromCallActionFactory<E : KtExpression>(
         analysisResult: AnalysisResult,
```
Comment (nikitabobko) [inline plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createCallable/CreateCallableFromCallActionFactory.kt:45]:
```
class A

fun test() {
    val a: Int = A.foo(2)
}
```
1. `Create extension function 'A.Companion.foo'`
2. Select `Any`
EXP: 
```
class A {
	companion object
}

fun test() {
	val a: Int = A.foo(2)
}

private fun Any.foo(i: Int): Int {
	TODO("Not yet implemented")
}
```
ACT:
```
class A 

fun test() {
	val a: Int = A.foo(2)
}

private fun Any.foo(i: Int): Int {
	TODO("Not yet implemented")
}
```

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createCallable/CreateCallableFromCallActionFactory.kt [hunk 2]
```diff
@@ -124,13 +133,15 @@ sealed class CreateCallableFromCallActionFactory<E : KtExpression>(
...
                 } else originalType
-                TypeInfo(finalType, Variance.IN_VARIANCE)
+                TypeInfo(finalType, Variance.IN_VARIANCE) to false
             }
             else -> throw AssertionError("Unexpected receiver: $receiver")
         }
     }
 
     protected fun getAbstractCallableInfo(mainCallable: CallableInfo, originalExpression: KtExpression): CallableInfo? {
+        if (mainCallable.isForCompanion) return null
```
Comment (nikitabobko) [inline plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createCallable/CreateCallableFromCallActionFactory.kt:143]: This `if` isn't covered with tests. If I comment it out and run tests then tests pass

**Review by nikitabobko — CHANGES_REQUESTED** [review]

**Review by nikitabobko — CHANGES_REQUESTED** [review]

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createCallable/CreateCallableFromCallActionFactory.kt [hunk 3]
```diff
@@ -140,8 +140,6 @@ sealed class CreateCallableFromCallActionFactory<E : KtExpression>(
     }
 
     protected fun getAbstractCallableInfo(mainCallable: CallableInfo, originalExpression: KtExpression): CallableInfo? {
-        if (mainCallable.isForCompanion) return null
-
```
[the remark this answers is not in the document]
Comment (t-kameyama) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createCallable/CreateCallableFromCallActionFactory.kt]: Reverted and added a test.

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createCallable/CreateCallableFromCallActionFactory.kt [hunk 4]
```diff
@@ -96,24 +97,32 @@ sealed class CreateCallableFromCallActionFactory<E : KtExpression>(
...
+                if (qualifierType != null) return TypeInfo(qualifierType, Variance.IN_VARIANCE) to false
 
                 if (receiver !is ClassQualifier) return null
                 val classifierType = receiver.descriptor.classValueType
-                if (classifierType != null) return TypeInfo(classifierType, Variance.IN_VARIANCE)
+                if (classifierType != null) return TypeInfo(classifierType, Variance.IN_VARIANCE) to false
 
-                val javaClassifier = receiver.descriptor as? JavaClassDescriptor ?: return null
+                val javaClassifier = receiver.descriptor as? JavaClassDescriptor
+                    ?: return TypeInfo(receiver.descriptor.defaultType, Variance.IN_VARIANCE) to true
```
Comment (nikitabobko) [inline plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createCallable/CreateCallableFromCallActionFactory.kt:122]: I just found `TypeInfo.StaticContextRequired`. It seems that the boolean flag `isForCompanion` you introduced is actually unnecessary and can be replaced with `org.jetbrains.kotlin.idea.quickfix.createFromUsage.callableBuilder.TypeInfo#getStaticContextRequired`

**Review by nikitabobko — CHANGES_REQUESTED** [review]

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createCallable/CreateCallableFromCallActionFactory.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (nikitabobko) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createCallable/CreateCallableFromCallActionFactory.kt]: `org.jetbrains.kotlin.idea.quickfix.QuickFixTestGenerated.CreateFromUsage.CreateFunction.Call.Abstract#testCompanion`  test doesn't fail anymore if you comment out this `if`.

It seems that after you migrated from `isForCompanion` to `staticContextRequired` this `if` isn't required anymore for some reason. Can you please elaborate why or cover this `if` with new tests?

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createVariable/CreateParameterFromUsageFix.kt [hunk 5]
```diff
@@ -92,6 +92,7 @@ class CreateParameterFromUsageFix<E : KtElement>(
             callableInfosFactory: (E) -> List<CallableInfo>?
         ): CreateParameterFromUsageFix<E> = CreateParameterFromUsageFix(element, dataProvider = fun(element): CreateParameterData<E>? {
             val info = callableInfosFactory.invoke(element)?.singleOrNull().safeAs<PropertyInfo>() ?: return null
+            if (info.receiverTypeInfo.staticContextRequired) return null
```
Comment (nikitabobko) [inline plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createVariable/CreateParameterFromUsageFix.kt:95]: I failed to find any test which covers this `if`. Are there any such tests? Or if this `if` isn't more required then can you please elaborate why it's not required (and this case let's remove this `if`) or if it's required then please add tests that cover this `if`

**Review by nikitabobko — CHANGES_REQUESTED** [review]
Let's resolve two comments and then I am happy to merge this PR

Please notice that for those two comments I am expecting:
1. either an explanation why after the migration from `isForCompanion` to `staticContextRequired` those `if`s are not required anymore.
2. or commits that cover those `if`s with tests.

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createCallable/CreateCallableFromCallActionFactory.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (t-kameyama) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createCallable/CreateCallableFromCallActionFactory.kt]: This `if` is not needed because `originalReceiverTypeInfo` becomes `TypeInfo.StaticContextRequired` and is returned here:

https://github.com/JetBrains/intellij-community/blob/aa444258814376186c42a5c41e57a6ff83320b49/plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createCallable/CreateCallableFromCallActionFactory.kt#L140

[the remark this answers is not in the document]
Comment (t-kameyama) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createCallable/CreateCallableFromCallActionFactory.kt]: I removed this `if`: aa44425

File: plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createVariable/CreateParameterFromUsageFix.kt (hunk 5 above)
Comment (t-kameyama) [reply plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/quickfix/createFromUsage/createVariable/CreateParameterFromUsageFix.kt:95]: If this `if` statement is removed, the following tests will fail.

- testData/quickfix/createFromUsage/createClass/referenceExpression/enumEntryWithQualifier.kt
- testData/quickfix/createFromUsage/createClass/referenceExpression/classWithQualifier.kt

**Review by nikitabobko — CHANGES_REQUESTED** [review]
Sorry, I hope that it's going to be the last review round trip. The Pull Request breaks those tests:
```
org.jetbrains.kotlin.idea.quickfix.QuickFixMultiFileTestGenerated$AddSpreadOperatorForArrayAsVarargAfterSam.testWithError
org.jetbrains.kotlin.idea.quickfix.QuickFixMultiFileTestGenerated$CreateFromUsage$CreateClass$ReferenceExpression.testEnumEntryWithJavaEnumQualifier
org.jetbrains.kotlin.idea.quickfix.QuickFixMultiFileTestGenerated$CreateFromUsage$CreateClass$ReferenceExpression.testEnumEntryWithJavaEnumSuperclass
org.jetbrains.kotlin.idea.quickfix.QuickFixMultiFileTestGenerated$CreateFromUsage$CreateClass$ReferenceExpression.testEnumEntryWithJavaNonEnumQualifier
org.jetbrains.kotlin.idea.quickfix.QuickFixMultiFileTestGenerated$CreateFromUsage$CreateClass$ReferenceExpression.testEnumEntryWithJavaNonEnumSuperclass
org.jetbrains.kotlin.idea.quickfix.QuickFixMultiFileTestGenerated$CreateFromUsage$CreateClass$ReferenceExpression.testObjectWithJavaQualifier
org.jetbrains.kotlin.idea.quickfix.QuickFixMultiFileTestGenerated$CreateFromUsage$CreateFunction$Call.testStaticExtensionFunOnJavaClass
org.jetbrains.kotlin.idea.quickfix.QuickFixMultiFileTestGenerated$CreateFromUsage$CreateVariable$PrimaryParameter.testStaticValOnJavaClass
org.jetbrains.kotlin.idea.quickfix.QuickFixMultiFileTestGenerated$CreateFromUsage$CreateVariable$Property.testStaticExtensionValOnJavaType
org.jetbrains.kotlin.idea.quickfix.QuickFixMultiFileTestGenerated$CreateFromUsage$CreateVariable$Property.testStaticValOnJavaClass
org.jetbrains.kotlin.idea.quickfix.QuickFixMultiFileTestGenerated$CreateFromUsage$CreateVariable$Property.testStaticValOnJavaInterface
org.jetbrains.kotlin.idea.quickfix.QuickFixMultiFileTestGenerated$CreateFromUsage$CreateVariable$Property.testStaticVarOnJavaInterface
org.jetbrains.kotlin.idea.quickfix.QuickFixTestGenerated$AutoImports.testImportEnumValueMissingReceiver
org.jetbrains.kotlin.idea.quickfix.QuickFixTestGenerated$AutoImports.testImportFunctionMissingReceiver
```

TLDR: don't suggest adding extension members on Java classes
